MekHQ/src/mekhq/campaign/universe/Systems.java

```diff
@@ -0,0 +1,500 @@
+        } catch (ParseException e) {
+            MekHQ.getLogger().error(getClass(), "initialize()", e); //$NON-NLS-1$
+        }
+    }
+    
+    private void done() {
+        initialized = true;
+        initializing = false;
+    }
+
+    public boolean isInitialized() {
+        return initialized;
+    }
+
+    private void updateSystems(FileInputStream source) {
+        final String METHOD_NAME = "updateSystems(FileInputStream)"; //$NON-NLS-1$
+        // JAXB unmarshaller closes the stream it doesn't own. Bad JAXB. BAD.
+        try(InputStream is = new FilterInputStream(source) {
+                @Override
+                public void close() {  }
+            }) {
+            // Reset the file stream
+            source.getChannel().position(0);
+
+            LocalSystemList systems = unmarshaller.unmarshal(
+                    MekHqXmlUtil.createSafeXmlSource(is), LocalSystemList.class).getValue();
+
+            // Run through the list again, this time creating and updating systems as we go
+            for( PlanetarySystem system : systems.list ) {
+                PlanetarySystem oldSystem = systemList.get(system.getId());
+                if( null == oldSystem ) {
+                    systemList.put(system.getId(), system);
+                } else {
+                    // Update with new data
+                    oldSystem.copyDataFrom(system);
+                    system = oldSystem;
+                }
+            }
+            
+            // Process system deletions
+            for( String systemId : systems.toDelete ) {
+                if( null != systemId ) {
+                    systemList.remove(systemId);
+                }
+            }
+        } catch (JAXBException e) {
+            MekHQ.getLogger().error(getClass(), METHOD_NAME, e);
+        } catch(IOException e) {
+            MekHQ.getLogger().error(getClass(), METHOD_NAME, e);
+        }
+    }
+    
+    private void generateSystems() throws DOMException, ParseException {
+        generateSystems("data/universe/planets", "data/universe/systems.xml");
+    }
+    
+    private void generateSystems(String planetsPath, String defaultFilePath) throws DOMException, ParseException {
+        final String METHOD_NAME = "generateSystems()"; //NON-NLS-1$
+        
+        MekHQ.getLogger().log(getClass(), METHOD_NAME, LogLevel.INFO,
+                "Starting load of system data from XML..."); //$NON-NLS-1$
+        long currentTime = System.currentTimeMillis();
+        synchronized (LOADING_LOCK) {
+            // Step 1: Initialize variables.
+            if( null == systemList ) {
+                systemList = new ConcurrentHashMap<>();
+            }
+            systemList.clear();
+            if( null == systemGrid ) {
+                systemGrid = new HashMap<>();
+            }
+            // Be nice to the garbage collector
+            for( Map.Entry<Integer, Map<Integer, Set<PlanetarySystem>>> systemGridColumn : systemGrid.entrySet() ) {
```

Comment:
Are any references held to the inner Map's and Set's of `systemGrid`? It doesn't appear so. If that's the case, you can just call `systemGrid.Clear()`, and that should ensure all of its elements are available for GC (without the added book-keeping of updating all of the various data structures).

Reply:
Its not my code. This was ported over from Planets.java. I am not sure why its done in the for-loop but I figured there must be a reason so I didn't change it.